fix functionalize() input metadata mutations, migrate core test to use functionalize()

bdhirsh opened the pull request:
Although this won't fully solve the "input metadata mutations" issue for AOTAutograd, there's no reason functionalization should break on input metadata mutations - this PR updates functionalization to use `as_strided_()` to mutate input metadata when necessary, similar to how it uses `copy_()` to handle input data mutations.

I also switched the functionalization tests over to using the proper `functionalize()` from functorch, which resulted in a few expect test changes.

Stack from [ghstack](https://github.com/ezyang/ghstack):
* #85844
* #85875
* #85681
* __->__ #82601
* #82602
* #82771
* #83265

bdhirsh committed:
fix functionalize() input metadata mutations, migrate core test to use functionalize()

[ghstack-poisoned]

bdhirsh committed:
Update on "fix functionalize() input metadata mutations, migrate core test to use functionalize()"




[ghstack-poisoned]

ezyang commented on test/test_functionalization.py:
```diff
@@ -3,9 +3,9 @@
-from torch.utils._pytree import tree_map
 from torch.fx.experimental.proxy_tensor import make_fx
 from torch.fx.passes.reinplace import reinplace
+from functorch.experimental import functionalize
```
since the build systems are not unified yet, ideally you should test if functorch is actually importable

bdhirsh replied:
oh, looks like CI is actually failing for that reason....

Any idea on the timeline for unifying build systems? Otherwise I guess I could try to import `functionalize()`, and fall back to the roll-your-own version that was in this file.

ezyang replied:
to test in CI, just update the CI to start building functorch after regular pytorch

@zou3519 knows unified build timeline

zou3519 replied:
A short fix is to install functorch in all CI configurations so that it's available for testing.

Unified build timeline: by 1.13, unless people have reasons for it earlier

ezyang reviewed (approved):
short and sweet (well maybe not the expect test updates lol)

bdhirsh committed:
Update on "fix functionalize() input metadata mutations, migrate core test to use functionalize()"

Although this won't fully solve the "input metadata mutations" issue for AOTAutograd, there's no reason functionalization should break on input metadata mutations - this PR updates functionalization to use `as_strided_()` to mutate input metadata when necessary, similar to how it uses `copy_()` to handle input data mutations.

I also switched the functionalization tests over to using the proper `functionalize()` from functorch, which resulted in a few expect test changes.




[ghstack-poisoned]

bdhirsh committed:
Update on "fix functionalize() input metadata mutations, migrate core test to use functionalize()"

Although this won't fully solve the "input metadata mutations" issue for AOTAutograd, there's no reason functionalization should break on input metadata mutations - this PR updates functionalization to use `as_strided_()` to mutate input metadata when necessary, similar to how it uses `copy_()` to handle input data mutations.

I also switched the functionalization tests over to using the proper `functionalize()` from functorch, which resulted in a few expect test changes.




[ghstack-poisoned]

bdhirsh committed:
Update on "fix functionalize() input metadata mutations, migrate core test to use functionalize()"

Although this won't fully solve the "input metadata mutations" issue for AOTAutograd, there's no reason functionalization should break on input metadata mutations - this PR updates functionalization to use `as_strided_()` to mutate input metadata when necessary, similar to how it uses `copy_()` to handle input data mutations.

I also switched the functionalization tests over to using the proper `functionalize()` from functorch, which resulted in a few expect test changes.




[ghstack-poisoned]

bdhirsh committed:
Update on "fix functionalize() input metadata mutations, migrate core test to use functionalize()"

Although this won't fully solve the "input metadata mutations" issue for AOTAutograd, there's no reason functionalization should break on input metadata mutations - this PR updates functionalization to use `as_strided_()` to mutate input metadata when necessary, similar to how it uses `copy_()` to handle input data mutations.

I also switched the functionalization tests over to using the proper `functionalize()` from functorch, which resulted in a few expect test changes.




[ghstack-poisoned]

bdhirsh committed:
Update on "fix functionalize() input metadata mutations, migrate core test to use functionalize()"

Although this won't fully solve the "input metadata mutations" issue for AOTAutograd, there's no reason functionalization should break on input metadata mutations - this PR updates functionalization to use `as_strided_()` to mutate input metadata when necessary, similar to how it uses `copy_()` to handle input data mutations.

I also switched the functionalization tests over to using the proper `functionalize()` from functorch, which resulted in a few expect test changes.




[ghstack-poisoned]

bdhirsh committed:
Update on "fix functionalize() input metadata mutations, migrate core test to use functionalize()"

Although this won't fully solve the "input metadata mutations" issue for AOTAutograd, there's no reason functionalization should break on input metadata mutations - this PR updates functionalization to use `as_strided_()` to mutate input metadata when necessary, similar to how it uses `copy_()` to handle input data mutations.

I also switched the functionalization tests over to using the proper `functionalize()` from functorch, which resulted in a few expect test changes.




[ghstack-poisoned]

bdhirsh committed:
Update on "fix functionalize() input metadata mutations, migrate core test to use functionalize()"

Although this won't fully solve the "input metadata mutations" issue for AOTAutograd, there's no reason functionalization should break on input metadata mutations - this PR updates functionalization to use `as_strided_()` to mutate input metadata when necessary, similar to how it uses `copy_()` to handle input data mutations.

I also switched the functionalization tests over to using the proper `functionalize()` from functorch, which resulted in a few expect test changes.




[ghstack-poisoned]

bdhirsh committed:
Update on "fix functionalize() input metadata mutations, migrate core test to use functionalize()"

Although this won't fully solve the "input metadata mutations" issue for AOTAutograd, there's no reason functionalization should break on input metadata mutations - this PR updates functionalization to use `as_strided_()` to mutate input metadata when necessary, similar to how it uses `copy_()` to handle input data mutations.

I also switched the functionalization tests over to using the proper `functionalize()` from functorch, which resulted in a few expect test changes.




[ghstack-poisoned]

bdhirsh committed:
Update on "fix functionalize() input metadata mutations, migrate core test to use functionalize()"

Although this won't fully solve the "input metadata mutations" issue for AOTAutograd, there's no reason functionalization should break on input metadata mutations - this PR updates functionalization to use `as_strided_()` to mutate input metadata when necessary, similar to how it uses `copy_()` to handle input data mutations.

I also switched the functionalization tests over to using the proper `functionalize()` from functorch, which resulted in a few expect test changes.




[ghstack-poisoned]

bdhirsh committed:
Update on "fix functionalize() input metadata mutations, migrate core test to use functionalize()"

Although this won't fully solve the "input metadata mutations" issue for AOTAutograd, there's no reason functionalization should break on input metadata mutations - this PR updates functionalization to use `as_strided_()` to mutate input metadata when necessary, similar to how it uses `copy_()` to handle input data mutations.

I also switched the functionalization tests over to using the proper `functionalize()` from functorch, which resulted in a few expect test changes.




[ghstack-poisoned]
